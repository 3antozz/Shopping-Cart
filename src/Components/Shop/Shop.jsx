import styles from "./Shop.module.css";
import ProductCard from "../Product-Card/Product-Card";
import PropTypes from "prop-types";
import { useOutletContext } from "react-router-dom";
import { useEffect, useState } from "react";
import { LoaderCircle, Search } from "lucide-react";
export default function Shop() {
    const { products, error, loading, handleAddToCart } = useOutletContext();
    const [popup, setPopup] = useState([]);
    const [inputValue, setInputValue] = useState("");
    const [filtered, setFiltered] = useState(products || []);

    useEffect(() => {
        if (products) {
            setFiltered(products)
        }
    }, [products]);

    function handleSearch (event) {
        setInputValue(event.target.value);
        if (event.target.value === "") {
            setFiltered(products);
        } else {
            const filteredProducts = products.filter((product) => product.title.toLowerCase().includes(event.target.value.toLowerCase()));
            setFiltered(filteredProducts);
        }
    }

    function handlePopup() {
        setPopup((prev) => [...prev, true]);
        setTimeout(() => {
            setPopup((prev) => {
                const curr = prev.slice();
                curr.pop();
                return curr;
            });
        }, 4000);
    }

    function preventSubmit (event) {
        event.preventDefault();
    }

    if (error)
        return (
            <div className={styles.error}>
                <h1>Oops, an Error has Occured! Please try again later</h1>
            </div>
        );
    if (loading)
        return (
            <div className={styles.loading}>
                <LoaderCircle
                    className={styles.spinner}
                    size={75}
                    color="rgba(255, 0, 0, 0.745)"
                    data-testid="loading-container"
                />
            </div>
        );
    return (
        <div className={styles.shop} data-testid="shop-container">
            <form className={styles.search} onSubmit={preventSubmit}>
                <label htmlFor="search"></label>
                <input
                    type="text"
                    id="search"
                    placeholder="Search for a product"
                    value={inputValue}
                    onChange={handleSearch}
                />
                <Search size={24} />
            </form>
            <div className={styles.products}>
                {popup.length > 0 && (
                    <div className={styles.popups}>
                        {popup.map((pop, index) => (
                            <div key={index} className={styles.popup}>
                                Item added to the cart!
                            </div>
                        ))}
                    </div>
                )}

                {filtered.map((product) => (
                    <ProductCard
                        key={product.id}
                        product={product}
                        onSubmit={handleAddToCart}
                        handlePopup={handlePopup}
                    />
                ))}
            </div>
        </div>
    );
}

Shop.propTypes = {
    products: PropTypes.object.isRequired,
};
