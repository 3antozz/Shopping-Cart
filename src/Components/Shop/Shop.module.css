.shop {
    display: flex;
    flex-wrap: wrap;
    padding: 2rem 8%;
    justify-content: center;
    gap: 5dvh;
    position: relative;
}

.popups {
    position: fixed;
    top: 25px;
    right: 10px;
    z-index: 999;
}

.popup {
    background-color: beige;
    margin-bottom: 1rem;
    padding: 1.5rem;
    border-radius: 15px;
    font-size: 1.3rem;
}