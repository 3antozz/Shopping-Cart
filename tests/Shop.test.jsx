import { describe, it, expect, beforeEach, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import Shop from "../src/Components/Shop/Shop.jsx";
import { MemoryRouter, Outlet, Routes, Route } from "react-router-dom";
import { act } from "react";
import userEvent from "@testing-library/user-event";

describe("Shop Component with products no loading", () => {
    const context = {
        products: [
            {
                id: 1,
                title: "test",
                price: 420,
                category: "...",
                description: "...",
                image: "...",
                quantity: 1,
            },
            {
                id: 2,
                title: "3antozz",
                price: 69,
                category: "...",
                description: "...",
                image: "...",
                quantity: 1,
            },
        ],
        error: false,
        loading: false,
        handleAddToCart: vi.fn(),
    };
    beforeEach(() => {
        act(() => {
            render(
                <MemoryRouter initialEntries={["/shop"]}>
                    <Routes>
                        <Route
                            path="/"
                            element={
                                <div>
                                    <h1>APP ROUTE</h1>
                                    <Outlet context={context} />
                                </div>
                            }>
                            <Route path="shop" element={<Shop />}></Route>
                        </Route>
                    </Routes>
                </MemoryRouter>
            );
        })
    });
    it("renders shop container", () => {
        expect(screen.getByTestId(/shop-container/i)).toBeInTheDocument();
    });
    it("renders product card container", () => {
        expect(screen.getAllByTestId("container").length).toBe(2);
    });
    it("renders search input", () => {
        expect(screen.getByPlaceholderText(/search for a product/i)).toBeInTheDocument();
    });
    it("should render popup when product added to card", async () => {
        const user = userEvent.setup();
        const cartButton = screen.getAllByRole("button", {name: /Add to Cart/i});

        await user.click(cartButton[0]);
        expect(screen.getByText("Item added to the cart!")).toBeInTheDocument();
    })
        it("matches snapshot", () => {
            const context = {
                products: [
                    {
                        id: 1,
                        title: "test",
                        price: 420,
                        category: "...",
                        description: "...",
                        image: "...",
                        quantity: 1,
                    },
                    {
                        id: 2,
                        title: "3antozz",
                        price: 69,
                        category: "...",
                        description: "...",
                        image: "...",
                        quantity: 1,
                    },
                ],
                error: false,
                loading: false,
                handleAddToCart: vi.fn(),
            };
            const { container } =         act(() => {
                render(
                    <MemoryRouter initialEntries={["/shop"]}>
                        <Routes>
                            <Route
                                path="/"
                                element={
                                    <div>
                                        <h1>APP ROUTE</h1>
                                        <Outlet context={context} />
                                    </div>
                                }>
                                <Route path="shop" element={<Shop />}></Route>
                            </Route>
                        </Routes>
                    </MemoryRouter>
                );
            })
            expect(container).toMatchSnapshot();
        });
});
describe("Shop component while loading", () => {
    const context = {
        products: [
            {
                id: 1,
                title: "test",
                price: 69,
                category: "...",
                description: "...",
                image: "...",
                quantity: 1,
            },
        ],
        error: false,
        loading: true,
        handleAddToCart: vi.fn(),
    };
    
    it("renders loading spinner", () => {
        render(
            <MemoryRouter initialEntries={["/shop"]}>
            <Routes>
                <Route
                    path="/"
                    element={
                        <div>
                            <h1>APP ROUTE</h1>
                            <Outlet context={context} />
                        </div>
                    }>
                    <Route path="shop" element={<Shop />}></Route>
                </Route>
            </Routes>
        </MemoryRouter>
        );
        expect(screen.getByTestId("loading-container")).toBeInTheDocument();
    })
})
describe("Shop component with an error", () => {
    const context = {
        products: [
            {
                id: 1,
                title: "test",
                price: 69,
                category: "...",
                description: "...",
                image: "...",
                quantity: 1,
            },
        ],
        error: true,
        loading: false,
        handleAddToCart: vi.fn(),
    };
    
    it("renders loading spinner", () => {
        render(
            <MemoryRouter initialEntries={["/shop"]}>
            <Routes>
                <Route
                    path="/"
                    element={
                        <div>
                            <h1>APP ROUTE</h1>
                            <Outlet context={context} />
                        </div>
                    }>
                    <Route path="shop" element={<Shop />}></Route>
                </Route>
            </Routes>
        </MemoryRouter>
        );
        expect(screen.getByText(/Oops, an Error has Occured! Please try again later/i)).toBeInTheDocument();
    })
})