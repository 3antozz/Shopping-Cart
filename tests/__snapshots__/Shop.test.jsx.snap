// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop Component with products no loading > matches snapshot 1`] = `undefined`;
